src/bit-fields-macros/src/utils.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt;
+
+/// Type used to ensure safety by avoiding some arithmetic operations.
+///
+/// Without this type we would parse and store the size of the bit field as a `u8`. The following
+/// operations would then use the generic `u8` implementations which the compiler cannot guarantee
+/// are safe. By using this enum to restrict the type we can avoid these operations. This is a
+/// small point, but helps a little.
+#[derive(Debug, Clone, Copy)]
+pub enum DataType {
+    /// `u8`
+    U8 = 8,
+    /// `u16`
+    U16 = 16,
+    /// `u32`
+    U32 = 32,
+    /// `u64`
+    U64 = 64,
+    /// `u128`
+    U128 = 128,
+}
+
+impl DataType {
+    // When const trait implementations are stabilized this should be
+    // `impl const From<DataType> for u8`.
+    /// Returns the size in bits of the data type.
+    #[allow(clippy::as_conversions)]
+    pub const fn size(self) -> u8 {
+        self as u8
+    }
+}
+
+impl fmt::Display for DataType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Self::U8 => write!(f, "u8"),
+            Self::U16 => write!(f, "u16"),
+            Self::U32 => write!(f, "u32"),
+            Self::U64 => write!(f, "u64"),
+            Self::U128 => write!(f, "u128"),
+        }
+    }
+}
+
+impl std::convert::TryFrom<&str> for DataType {
+    type Error = DataTypeTryFromError;
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        match s {
+            "u8" => Ok(Self::U8),
+            "u16" => Ok(Self::U16),
+            "u32" => Ok(Self::U32),
+            "u64" => Ok(Self::U64),
+            "u128" => Ok(Self::U128),
+            _ => Err(DataTypeTryFromError),
+        }
+    }
+}
+
+/// Error type for `impl TryFrom<&str> for DataType`
+#[derive(Debug, PartialEq, Eq)]
+pub struct DataTypeTryFromError;
```

Comment:
I believe `thiserror` does not work for unit structs.

Reply:
From the [thiserror documentation](https://docs.rs/thiserror/latest/thiserror/#details): "Errors may be enums, structs with named fields, tuple structs, **or unit structs**."

Reply:
I think their documentation may be wrong. A unit struct contains no fields (e.g. `struct MyStruct;`) where as a newtype struct contains a single field (e.g. `struct MyStruct(u32);`)

I believe they mean a new-type struct since in their documentation this is the example they give:
```rust
#[derive(Error, Debug)]
#[error(transparent)]
pub struct PublicError(#[from] ErrorRepr);
```
and they give no example of a unit struct.

Unless I'm missing something? (may be worth opening PR to fix their documentation if this is the case).

---

https://doc.rust-lang.org/rust-by-example/custom_types/structs.html

> Unit structs, which are field-less, are useful for generics.

https://doc.rust-lang.org/rust-by-example/generics/new_types.html

> struct Years(i64);

Reply:
The documentation states "A Display impl is generated for your error if you provide #[error("...")] messages **on the struct** or each variant of your enum, as shown above in the example."

I just verified that the following works:

```rs
/// Error type for `impl TryFrom<&str> for DataType`
#[derive(Debug, PartialEq, Eq, Error)]
#[error("Bad backing data type given. Only accepted types are `u8`, `u16`, `u32`, `u64` and \
             `u128`.")]
pub struct DataTypeTryFromError;
```

However, I also noticed that `thiserror` is not actually a dependency of `bit-fields-macros`.

Reply:
Implemented this usage of `thiserror`.